AmplifyPlugins/Auth/AWSAuthPlugin/Dependency/AuthorizationProviderAdapter+SignedInSession.swift

```diff
@@ -0,0 +1,194 @@
+
+        // User can be signed in to user pools or federated to identityPool.
+        // Call getTokens first to verify that the signedIn user is in UserPool or not.
+        awsMobileClient.getTokens { [weak self] tokens, error in
+            guard let self = self else {
+                return
+            }
+            guard error == nil else {
+
+                if let urlError = error as NSError?, urlError.domain == NSURLErrorDomain {
+                    let authSession = AuthCognitoSignedInSessionHelper.offlineSignedInSession()
+                    completionHandler(.success(authSession))
+                } else if let awsMobileClientError = error as? AWSMobileClientError {
+
+                    if case .notSignedIn = awsMobileClientError {
+                        // The user is signed in using Cognito Identity Pool, in this case we just
+                        // fetch the aws credentials and identity id.
+                        let tokensError = AmplifyAuthError.service(
+                            AuthPluginErrorConstants.cognitoTokenSignedInThroughCIDPError.errorDescription,
+                            AuthPluginErrorConstants.cognitoTokenSignedInThroughCIDPError.recoverySuggestion,
+                            AWSCognitoAuthError.invalidAccountTypeException)
+                        let userSubError = AmplifyAuthError.service(
+                            AuthPluginErrorConstants.userSubSignedInThroughCIDPError.errorDescription,
+                            AuthPluginErrorConstants.userSubSignedInThroughCIDPError.recoverySuggestion,
+                            AWSCognitoAuthError.invalidAccountTypeException)
+                        self.retrieveAWSCredentials(withTokensResult: .failure(tokensError),
+                                                    userSubResult: .failure(userSubError),
+                                                    completionHandler: completionHandler)
+
+                    } else if case .unableToSignIn = awsMobileClientError {
+                        // Refresh token has expired, the user need to signIn again. This error is thrown from
+                        // awsmobileclient signOut or releaseSignInWait.
+                        let authSession = AuthCognitoSignedInSessionHelper.expiredSignedInSession()
+                        completionHandler(.success(authSession))
+                    }
+                }
+                let authError = AuthErrorHelper.toAmplifyAuthError(error!)
+                let authSession = AuthCognitoSignedInSessionHelper.signedInSession(withUnhandledError: authError)
+                completionHandler(.success(authSession))
+                return
+            }
+
+            guard let tokens = tokens else {
+                // This should not happen, return an unknown error.
+                let error = AmplifyAuthError.unknown("""
+                    Could not fetch the cognito tokens
```

Comment:
Put some more context here: we reported the state/session/user as signedIn, but we can't retrieve Cognito Identity credentials, or user pool tokens--describe the situation more fully. I'm not saying we'll ever hit this error, but when we do, we can save ourselves a lot of effort if we clearly identify the expectations going into it.

Reply:
Added, let me know if the message is ok..